[mlir][vector] Prevent folding non memref-type gather into maskedload

sagarkulkarni19 opened the pull request:
This patch fixes an issue in the FoldContiguousGather pattern which was incorrectly folding vector.gather operations with contiguous indices into vector.maskedload operations regardless of the base operand type.

While vector.gather operations can work on both tensor and memref types, vector.maskedload operations are only valid for memref types. The pattern was incorrectly lowering a tensor-based gather into a masked-load, which is invalid.

This fix adds a type check to ensure the pattern only applies to memref-based gather operations.

sagar-rain committed:
[mlir][vector] Prevent folding non memref-type gather into maskedload

This patch fixes an issue in the FoldContiguousGather pattern which was
incorrectly folding vector.gather operations with contiguous indices into
vector.maskedload operations regardless of the base operand type.

While vector.gather operations can work on both tensor and memref types,
vector.maskedload operations are only valid for memref types. The pattern
was incorrectly lowering a tensor-based gather into a masked-load, which
is invalid.

This fix adds a type check to ensure the pattern only applies to memref-based
gather operations.

sagarkulkarni19 commented:
@Hardcode84 @ashay

Hardcode84 reviewed (commented):
This is a weird discrepancy between vector ops I wasn't aware of. Fix itself is LGTM.

Please also wait for @banach-space or @dcaballe review.

banach-space commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5340,6 +5340,9 @@ class FoldContiguousGather final : public OpRewritePattern<GatherOp> {
   LogicalResult matchAndRewrite(GatherOp op,
                                 PatternRewriter &rewriter) const override {
+    if (!op.getBase().getType().isa<MemRefType>())
+      return failure();
```
[nit] Could you use `notifyMatchFailure`? Thanks!

sagarkulkarni19 replied:
Good point, done.

banach-space commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -3149,6 +3149,18 @@ func.func @contiguous_gather_step(%base: memref<?xf32>,
 // -----
 
+// CHECK-LABEL: @dont_fold_tensor_type_contiguous_gather
+func.func @dont_fold_tensor_type_contiguous_gather(%base: tensor<8xf32>, %mask: vector<4xi1>, %pass_thru: vector<4xf32>) -> vector<4xf32> {
```
1. `no_fold` is shorter than `dont_fold` :)
2. The naming format seems to be `@contiguous_gather_{other_stuff}`, so `@dont_fold_tensor_type_contiguous_gather` -> `@no_fold_contiguous_gather_tensor` (or something similar)

sagarkulkarni19 replied:
Changed it to `@no_fold_contiguous_gather_tensor`.

banach-space commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -3149,6 +3149,18 @@ func.func @contiguous_gather_step(%base: memref<?xf32>,
+  %indices = arith.constant dense<[0, 1, 2, 3]> : vector<4xindex>
+  // CHECK: vector.gather
+  // CHECK-NOT: vector.maskedload
+  %0 = vector.gather %base[%c0][%indices], %mask, %pass_thru : tensor<8xf32>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
```
[nit] Would you mind splitting into two lines like we did above?

sagarkulkarni19 replied:
Done.

banach-space reviewed (approved):
The fix makes sense to me — thanks for taking care of it! I've left a few nits, but otherwise LGTM.

@dcaballe and I were actually just talking about the “duality” of read/write ops in the Vector dialect — we agree it's something we should clean up and this PR confirms that.

Hardcode84 commented:
Side comment: gather takes `AnyShaped` so gather "loading" from vector will be accepted as well, which was probably never intended.

sagarkulkarni19 committed:
Merge branch 'main' into fix-fold-contiguous-gather-pattern

sagarkulkarni19 commented:
@banach-space I've addressed the comments. Lmk if it looks good.

@Hardcode84 @banach-space Can one of you help me merge it in since I do not have merge permission. Thanks!
